aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -226,35 +226,23 @@ impl<'r, 'l> SessionExt<'r, 'l> {
             }
 
             for (resource_tag, resources) in resource_groups {
-                let source_data = remote
-                    .get_resource_group_data(&addr, &resource_tag)
-                    .map_err(|_| common_error.clone())?;
-                let (mut source_data, create) = if let Some(source_data) = source_data {
-                    let source_data =
-                        bcs::from_bytes(&source_data).map_err(|_| common_error.clone())?;
-                    (source_data, false)
-                } else {
-                    (BTreeMap::new(), true)
-                };
+                let mut create = false;
+                let mut source_data = remote
+                    .get_remaining_resource_group(&addr, &resource_tag)
+                    .unwrap_or_else(|| {
+                        create = true;
+                        BTreeMap::new()
+                    });
 
                 for (struct_tag, current_op) in resources.into_resources() {
+                    if source_data.contains_key(&struct_tag) {
+                        return Err(common_error);
+                    }
                     match current_op {
-                        MoveStorageOp::Delete => {
-                            source_data
-                                .remove(&struct_tag)
-                                .ok_or_else(|| common_error.clone())?;
-                        },
-                        MoveStorageOp::Modify(new_data) => {
-                            let data = source_data
-                                .get_mut(&struct_tag)
-                                .ok_or_else(|| common_error.clone())?;
-                            *data = new_data;
-                        },
-                        MoveStorageOp::New(data) => {
+                        MoveStorageOp::Delete => {},
+                        MoveStorageOp::Modify(data) | MoveStorageOp::New(data) => {
                             let data = source_data.insert(struct_tag, data);
-                            if data.is_some() {
-                                return Err(common_error);
-                            }
+                            debug_assert!(data.is_none());
```

Comment:
this takes the assumption of parallel writes, which seems to be a dangerous assumption. why should the resolver know about blockstm? and what happens if blockstm changes and this becomes a real invariant violation?

if we wanted to just move fast here we could delete the error message, but I think that's also not quite right.

I believe we should figure out how to embrace clean, modular coding rather than let implementation details seep in.

Reply:
I rewrote some of it. It's now similar to original just no rereads